### PR TITLE
Release/0.4.0

### PR DESCRIPTION
**Pull request by sibartlett**
* Private and password-protected rooms
* Private 1-to-1 chat between XMPP users
* Giphy image search
* @all mentions are highlighted for everyone
* Basic server-side i18n support

### CODE REVIEW

**Comment by funkaoshi** [comment]
Ah yeah! Let's do this.

**Comment by hhaidar** [comment]
@all feature is already in master :P

**Comment by sibartlett** [comment]
@hhaidar it's not in the changelog on master though, so put it here

**Comment by franciscod** [comment]
nice!

File: media/js/views/modals.js [hunk 1]
```diff
@@ -176,4 +201,67 @@
...
+            this.$el.find('.giphy-results ul').empty();
+            this.$('.search-giphy').val('').focus();
+        },
+        stopReturn: function(e) {
+            if(e.keyCode === 13) {
+                return false;
+            }
+        },
+        loadGifs: _.throttle(function() {
+            console.log(1)
```
Comment (TimWolla) [inline media/js/views/modals.js]: Debug code.

File: media/js/views/modals.js [hunk 2]
```diff
@@ -176,4 +201,67 @@
...
+        refresh: function() {
+            this.$el.find('.giphy-results ul').empty();
+            this.$('.search-giphy').val('').focus();
+        },
+        stopReturn: function(e) {
+            if(e.keyCode === 13) {
+                return false;
+            }
+        },
+        loadGifs: _.throttle(function() {
```
Comment (TimWolla) [inline media/js/views/modals.js]: Throttle generates lots of unnecessary traffic, as the requests for the old images are not aborted and the GIFs tend to be rather big. Consider using debounce to hit the API when the user has stopped typing.

File: media/js/views/modals.js (hunk 1 above)
Comment (sibartlett) [inline media/js/views/modals.js]: Thank, the debug code has been removed.

File: media/js/views/modals.js (hunk 2 above)
Comment (sibartlett) [inline media/js/views/modals.js]: I've changed the code to use debounce.

**Comment by geoffwhittington** [comment]
Tests are passing: :+1:  :+1:

**Comment by sibartlett** [comment]
@geoffwhittington what tests? :laughing:

**Comment by rockneverdies55** [comment]
Is there any ETA for this to be merged in?

**Comment by tonghuix** [comment]
I test it OK, waiting for the release!

**Comment by sibartlett** [comment]
We hope to have this merged soon. We were waiting to release, so we could do some thorough testing. Also I think we're going to add some i18n support to this release.

**Comment by WhiteHatTux** [comment]
Can't wait for the new release. Is there anything special you would like to have tested?

**Pull request closed without merging**